### PR TITLE
[tutasdk] sdk should be able to make service calls

### PR DESCRIPTION
**Pull request by tuta-sudipg**
* NOTE: Compressed String not yet supported!
* RustGenerator should produce service files
* EntityFacade should be able to encrypt entities
* CreateSession ( without offline login and 2FA )
* NativeRestClient based on `hyper` to use in non-mobile ( desktop ) devices

### Test Notes
* regression test mail notifications for ios + android

### CODE REVIEW

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs [hunk 1]
```diff
@@ -41,15 +62,238 @@ pub trait EntityFacade: Send + Sync {
...
+					.get_type_model(dependency, association.ref_type)
+					.unwrap_or_else(|| {
+						panic!(
+							"Cannot find type model for: {:?}",
+							(dependency, association.ref_type)
+						)
+					});
+				let aggregation = association;
+				let instance_association = instance.get(&association_name.to_string()).unwrap();
+				if aggregation.cardinality == Cardinality::ZeroOrOne
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/entities/entity_facade.rs:248]: I would feel slightly better if this `if` was a `match` so that we explicitly mention all the cases. It might also be more readable.

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs [hunk 2]
```diff
@@ -41,15 +62,238 @@ pub trait EntityFacade: Send + Sync {
...
+					let parsed_entity = self.encrypt_and_map_to_literal_with_iv(
+						aggregated_type_model,
+						&instance_association.assert_dict(),
+						sk.clone(),
+						random_iv_provider.clone(),
+					)?;
+					let encrypted_aggregate = ElementValue::Dict(parsed_entity);
+					encrypted.insert(association_name.to_string(), encrypted_aggregate);
+				}
+			} else {
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/entities/entity_facade.rs:281]: I think this `if-else` is too big

File: tuta-sdk/rust/sdk/src/entity_client.rs [hunk 3]
```diff
@@ -23,6 +23,12 @@ pub struct EntityClient {
 	type_model_provider: Arc<TypeModelProvider>,
 }
 
+impl EntityClient {
+	pub fn get_headers_provider(&self) -> &Arc<HeadersProvider> {
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/entity_client.rs:27]: why is this necessary? Why is it reference to `Arc`?

File: tuta-sdk/rust/sdk/src/instance_mapper.rs [hunk 4]
```diff
@@ -564,36 +563,38 @@ impl Serializer for ElementValueSerializer {
 		Ok(ElementValue::Bool(v))
 	}
 
-	fn serialize_i8(self, _: i8) -> Result<Self::Ok, Self::Error> {
-		unsupported("i8")
+	fn serialize_i8(self, v: i8) -> Result<Self::Ok, Self::Error> {
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/instance_mapper.rs:566]: we don't have most number sized. I'm not sure if we should write code for them or leave them explicitly unsupported.

File: tuta-sdk/rust/sdk/src/lib.rs [hunk 5]
```diff
@@ -54,15 +63,21 @@ mod logging;
...
+pub mod services;
 mod simple_crypto;
-mod tutanota_constants;
+pub mod tutanota_constants;
 mod type_model_provider;
 mod typed_entity_client;
 mod user_facade;
 mod util;
 
+pub static CLIENT_VERSION: &str = env!("CARGO_PKG_VERSION");
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/lib.rs:79]: we've been passing client version from the users of the SDK. Why is it necessary?

File: tuta-sdk/rust/sdk/src/metamodel.rs [hunk 6]
```diff
@@ -34,6 +36,21 @@ pub enum ValueType {
...
+impl ValueType {
+	pub fn get_default(&self) -> ElementValue {
+		match self {
+			ValueType::String | ValueType::CompressedString => ElementValue::String(String::new()),
+			ValueType::Number => ElementValue::Number(0),
+			ValueType::Bytes => ElementValue::Bytes(Vec::new()),
+			ValueType::Date => ElementValue::Date(DateTime::default()),
+			ValueType::Boolean => ElementValue::Bool(false),
+			ValueType::GeneratedId | ValueType::CustomId => {
+				panic!("{self:?} is not a valid value type")
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/metamodel.rs:48]: error message is misleading

File: tuta-sdk/rust/sdk/src/services/generator.rs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	($x:expr, $res:expr, ()) => {
+		return Ok(());
+	};
+	($x:expr, $res:expr, $i: tt) => {
+		return $x.handle_response::<Self::Output>($res).await;
+	};
+}
+
+#[macro_export]
+macro_rules! service_impl {
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/services/generator.rs:32]: these macros looks a bit cryptic to me. I do not understand what `base` is. There's also no documentation on them either. Would be nice to document them (at least the interface but maybe also the impl) for those of us who are not so versed in arcane arts like me :sweat_smile:  🔮

Comment (charlag) [reply tuta-sdk/rust/sdk/src/services/generator.rs:32]: It's also two macros with the same name? Should they have different names perhaps?

File: tuta-sdk/rust/sdk/src/services/service_executor.rs [hunk 8]
```diff
@@ -0,0 +1,650 @@
...
+};
+use crate::type_model_provider::TypeModelProvider;
+use crate::{ApiCallError, HeadersProvider};
+use serde::Deserialize;
+use serde::Serialize;
+use std::sync::Arc;
+
+pub struct ServiceExecutor {
+	auth_headers_provider: Arc<HeadersProvider>,
+	crypto_facade: Arc<CryptoFacade>,
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/services/service_executor.rs:23]: I think this `CryptoFacade` here is the reason why we don't have strong guarantees about logged in state.

I think we should have similar separation as with `EntityClient` and have one without encryption support and one with encryption support.

File: tuta-sdk/rust/sdk/src/services/service_executor.rs [hunk 9]
```diff
@@ -0,0 +1,650 @@
...
+				})?;
+			let input_type_ref = I::type_ref();
+			let type_model = self
+				.type_model_provider
+				.get_type_model(input_type_ref.app, input_type_ref.type_)
+				.ok_or(ApiCallError::internal(format!(
+					"type {:?} does not exist",
+					input_type_ref
+				)))?;
+			let encrypted_parsed_entity = self.entity_facade.encrypt_and_map_to_literal(
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/services/service_executor.rs:137]: I think we should not call encrypt for entities that are not encrypted, like I mentioned before

File: tuta-sdk/rust/sdk/src/services/service_executor.rs [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+						.parse_entity::<OutputType>(decrypted_entity)
+						.map_err(|e| {
+							ApiCallError::internal_with_err(
+								e,
+								"Failed to parse encrypted entity into proper types",
+							)
+						})?;
+					Ok(typed_entity)
+				},
+				// `resolve_session_key()` only returns none if the entity is unencrypted, so
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/services/service_executor.rs:220]: should we match or just unwrap then?

File: tuta-sdk/rust/sdk/src/net/native_rest_client.rs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+use hyper::body::Incoming;
+use hyper::http::HeaderValue;
+use hyper::{HeaderMap, Request, Response};
+use hyper_rustls::HttpsConnector;
+use hyper_util::client::legacy::connect::HttpConnector;
+use hyper_util::client::legacy::Client;
+use hyper_util::rt::TokioExecutor;
+use std::collections::HashMap;
+
+pub struct NativeRestClient {
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/net/native_rest_client.rs:15]: I still think `reqwest` impl is much shorter but if it works, it works

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs [hunk 12]
```diff
@@ -41,15 +62,238 @@ pub trait EntityFacade: Send + Sync {
 		entity: ParsedEntity,
 		resolved_session_key: ResolvedSessionKey,
 	) -> Result<ParsedEntity, ApiCallError>;
+
+	fn encrypt_and_map_to_literal(
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/entities/entity_facade.rs:66]: I don't know if the "literal" in the name makes sense anymore, that was kind of js-speak for "object" I guess

File: tuta-sdk/rust/sdk/Cargo.toml [hunk 13]
```diff
@@ -27,11 +27,22 @@ rsa = "0.9.6"
...
 log = "0.4.22"
 simple_logger = "5.0.0"
 uniffi = { git = "https://github.com/mozilla/uniffi-rs.git", rev = "13a1c559cb3708eeca40dcf95dc8b3ccccf3b88c" }
 
+
+# dependency required for using mocks
+mockall = { version = "0.13.0", optional = true }
+mockall_double = { version = "0.3.1", optional = true }
+
+# only used for the native rest client
```
[the remark this answers is not in the document]
Comment (ganthern) [reply tuta-sdk/rust/sdk/Cargo.toml:39]: not sure what you mean, are the tests run 4 times?

File: tuta-sdk/rust/sdk/src/crypto/crypto_facade.rs [hunk 14]
```diff
@@ -32,22 +33,24 @@ const BUCKET_KEY_FIELD: &str = "bucketKey";
 
 #[derive(uniffi::Object)]
 pub struct CryptoFacade {
-	key_loader_facade: Arc<KeyLoaderFacade>,
+	key_loader_facade: Option<Arc<KeyLoaderFacade>>,
```
[the remark this answers is not in the document]
Comment (ganthern) [reply tuta-sdk/rust/sdk/src/crypto/crypto_facade.rs:36]: I also don't really like it. I think we can construct a proof that makes it impossible to call the methods requiring the KeyloaderFacade unless you successfully logged in :thinking:

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs [hunk 15]
```diff
@@ -41,15 +62,238 @@ pub trait EntityFacade: Send + Sync {
...
+			}
+		}
+		false
+	}
+
+	fn encrypt_value(
+		value_name: &str,
+		model_value: &ModelValue,
+		instance_value: &ElementValue,
+		session_key: &Option<GenericAesKey>,
```
[the remark this answers is not in the document]
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/entities/entity_facade.rs:101]: agreed...

File: tuta-sdk/rust/sdk/src/crypto/crypto_facade.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (charlag) [reply tuta-sdk/rust/sdk/src/crypto/crypto_facade.rs:36]: It is one thing to do unauthenticated requests (that is kinda hidden in headers provider) but another to request encrypted things. We have multiple `EntityClient`s partially for this reason.

I really think you need two top-level `ServiceExecutor`s, maybe they both delegate to the same thing under the hood though.

File: tuta-sdk/rust/sdk/src/crypto_entity_client.rs [hunk 16]
```diff
@@ -18,6 +18,12 @@ pub struct CryptoEntityClient {
 	instance_mapper: Arc<InstanceMapper>,
 }
 
+impl CryptoEntityClient {
+	pub fn get_crypto_facade(&self) -> &Arc<CryptoFacade> {
```
[the remark this answers is not in the document]
Comment (charlag) [reply tuta-sdk/rust/sdk/src/crypto_entity_client.rs:22]: I understand how it works on a technical level, but it feels really odd to do this.

File: tuta-sdk/rust/sdk/src/instance_mapper.rs (hunk 4 above)
Comment (charlag) [reply tuta-sdk/rust/sdk/src/instance_mapper.rs:566]: sorry, I meant we don't support most numbers

File: tuta-sdk/rust/sdk/src/entity_client.rs (hunk 3 above)
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/entity_client.rs:27]: Arc

File: tuta-sdk/rust/sdk/Cargo.toml (hunk 13 above)
[the remark this answers is not in the document]
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/Cargo.toml:39]: Tried to play with mockall.
But now, they are referenced only through test. So they can stay in dev-dependencies now.

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs [hunk 17]
```diff
@@ -41,15 +62,238 @@ pub trait EntityFacade: Send + Sync {
...
+					instance_value,
+					&sk,
+					random_iv_provider.clone(),
+				)?
+			}
+			encrypted.insert(key.to_string(), encrypted_value);
+		}
+
+		if type_model.element_type == ElementType::Aggregated && !encrypted.contains_key("_id") {
+			let randomizer = RandomizerFacade::from_core(rand_core::OsRng);
```
[the remark this answers is not in the document]
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/entities/entity_facade.rs]: injected a randomizer in entityFacade itself

File: tuta-sdk/rust/sdk/src/crypto_entity_client.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/crypto_entity_client.rs]: The whole get_* ( added in this commit ) is not needed

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/entities/entity_facade.rs:86]: changed the function name instead

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs (hunk 1 above)
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/entities/entity_facade.rs]: 🙇‍♂️

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs (hunk 2 above)
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/entities/entity_facade.rs]: Splitted the association == Aggregated to seperate function now

File: tuta-sdk/rust/sdk/src/instance_mapper.rs (hunk 4 above)
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/instance_mapper.rs]: True. maybe because we were using some random test Entity with `u8/i8` field. should not be part of this commit anyway

File: tuta-sdk/rust/sdk/src/crypto_entity_client.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/crypto_entity_client.rs]: Were mistakes. Should have not been committed

File: tuta-sdk/rust/sdk/src/lib.rs (hunk 5 above)
Comment (ganthern) [reply tuta-sdk/rust/sdk/src/lib.rs:79]: I felt adding it to the bump-version script (forgot to do it in the end) and having a compile-time constant is better than passing around a string that could be anything. 
It's not like we're in any danger of following SemVer with this and the SDK should probably have some idea which client version it's supposed to be compatible with, even if looked at in isolation.

Comment (charlag) [reply tuta-sdk/rust/sdk/src/lib.rs:79]: I'm not opposed to it but let's make it a separate change then?

Comment (charlag) [reply tuta-sdk/rust/sdk/src/lib.rs:79]: Also, there are cases where we want app version number and there are cases where it's more protocol-related. I'm not sure if we maybe should split them somehow.

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs [hunk 18]
```diff
@@ -232,66 +210,80 @@ impl EntityFacadeImpl {
...
+					"Cannot find type model for: {:?}",
+					(dependency, association.ref_type)
+				)
+			});
+		let instance_association = instance.get(&association_name.to_string()).unwrap();
+
+		match (&association.cardinality, instance_association) {
+			(Cardinality::ZeroOrOne, ElementValue::Null) => Ok(ElementValue::Null),
+			(_, ElementValue::Null) => {
+				panic!("Undefined attribute {}:{association_name}", type_model.name)
```
Comment (charlag) [inline tuta-sdk/rust/sdk/src/entities/entity_facade.rs:261]: when we can't find association it is `Err` but here we `panic()`. We should be consistent.

File: tuta-sdk/rust/sdk/src/lib.rs (hunk 5 above)
Comment (ganthern) [reply tuta-sdk/rust/sdk/src/lib.rs:79]: it's a separate commit now. 

In my mind I guess the major version is the protocol-related part while the other parts of the version are more about which client you're using. I'm sure we could improve on the current headers to make it more independent.

File: buildSrc/RustGenerator.js [hunk 19]
```diff
@@ -13,7 +13,7 @@ import { AssociationType, Type } from "../src/common/api/common/EntityConstants.
  */
 export function generateRustType({ type, modelName }) {
 	let typeName = mapTypeName(type.name, modelName)
-	let buf = `#[derive(uniffi::Record, Clone, Serialize, Deserialize, Debug)]
+	let buf = `#[derive(uniffi::Record, Clone, Serialize, Deserialize, Debug, PartialEq)]
```
[the remark this answers is not in the document]
Comment (tuta-sudipg) [reply buildSrc/RustGenerator.js]: done

File: tuta-sdk/rust/sdk/src/entities/entity_facade.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/entities/entity_facade.rs]: ended up checking the encrypted flag in the service executor

File: tuta-sdk/rust/sdk/src/entity_client.rs (hunk 3 above)
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/entity_client.rs]: should be fixed, removed the impl block

File: tuta-sdk/rust/sdk/src/metamodel.rs (hunk 6 above)
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/metamodel.rs:48]: Done

File: tuta-sdk/rust/sdk/src/crypto/crypto_facade.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ganthern) [reply tuta-sdk/rust/sdk/src/crypto/crypto_facade.rs:36]: we had a long think about this and we came up with a few ideas to prevent mistakes at compile time, but nothing that we would implement now since it would require some more work.

We did a version that wraps a service executor that can resolve keys around a "normal" service executor though to not damage any invariants that were there before.

File: tuta-sdk/rust/sdk/src/services/service_executor.rs [hunk 20]
```diff
@@ -0,0 +1,653 @@
...
+	auth_headers_provider: Arc<HeadersProvider>,
+	crypto_facade: Arc<CryptoFacade>,
+	entity_facade: Arc<dyn EntityFacade>,
+	instance_mapper: Arc<InstanceMapper>,
+	json_serializer: Arc<JsonSerializer>,
+	rest_client: Arc<dyn RestClient>,
+	type_model_provider: Arc<TypeModelProvider>,
+	base_url: String,
+}
+impl ServiceExecutor {
```
[the remark this answers is not in the document]
Comment (ganthern) [reply tuta-sdk/rust/sdk/src/services/service_executor.rs:70]: alright!

File: tuta-sdk/rust/sdk/src/services/generator.rs (hunk 7 above)
Comment (tuta-sudipg) [reply tuta-sdk/rust/sdk/src/services/generator.rs:77]: There is now further explanitation on how this macro is supposed to be used. doc is marked as:
```
```ignored
```
Without `ignored` cargo test will fail,
Since this macro expansion depends on `::services::service_executor::hidden` module which is not public.
